### ISSUE TITLE
Flickering and rendering issues (intel)

### ISSUE THREAD

**Issue by jeLee6gi** [yshui/compton#34]
### Platform
Archlinux 4.18.16-arch1-1-ARCH

### GPU, drivers, and screen setup
Intel HD 5500, xf86-video-intel 1:2.99.917+853+g0932a6b3-1, mesa 18.2.3-1

### Environment
qtile-python3

### Compton version
v3-rc2-18-gc0d7f9d

### Compton configuration:
```
backend = "glx";
glx-no-stencil = true;
glx-copy-from-front = false;
glx-use-copysubbuffermesa = false;
glx-no-rebind-pixmap = false;
glx-swap-method = "undefined";

shadow = false;
no-dnd-shadow = false;
no-dock-shadow = true;
clear-shadow = true;
shadow-radius = 0;
shadow-offset-x = -5;
shadow-offset-y = -5;
shadow-opacity = 0;
shadow-exclude = [
    "! name~=''",
    "name = 'Notification'",
    "name = 'Plank'",
    "name = 'Docky'",
    "name = 'Kupfer'",
    "name = 'xfce4-notifyd'",
    "name *= 'VLC'",
    "name *= 'compton'",
    "name *= 'Chromium'",
    "name *= 'Chrome'",
    "class_g = 'Conky'",
    "class_g = 'Kupfer'",
    "class_g = 'Synapse'",
    "class_g ?= 'Notify-osd'",
    "class_g ?= 'Cairo-dock'",
    "class_g ?= 'Xfce4-notifyd'",
    "class_g ?= 'Xfce4-power-manager'",
    "_GTK_FRAME_EXTENTS@:c"
];
shadow-ignore-shaped = false;

menu-opacity = 1;
inactive-opacity = 1;
active-opacity = 1;
frame-opacity = 1;
inactive-opacity-override = false;
alpha-step = 0.06;
blur-background-fixed = false;
blur-background-exclude = [
    "window_type = 'dock'",
    "window_type = 'desktop'"
];

fading = true;
fade-delta = 5
fade-in-step = 0.05;
fade-out-step = 0.05;
fade-exclude = [ ];

mark-wmwin-focused = true;
mark-ovredir-focused = true;
use-ewmh-active-win = true;
detect-rounded-corners = true;
detect-client-opacity = true;
refresh-rate = 0;

vsync = "opengl";
dbe = false;
paint-on-overlay = true;
sw-opti = false;
unredir-if-possible = false;
focus-exclude = [ ];
detect-transient = true;
detect-client-leader = true;

wintypes:
{
    tooltip =
    {
        fade = true;
        shadow = false;
        opacity = 0.85;
        focus = true;
    };
};
```

### Steps of reproduction
1. Run compton v3 with my config

### Expected behavior
Compton fades windows in and out and does vsyncing

### Current Behavior
The screen flickers and parts of windows are not redrawn.

### Other details
[Here is a video showing the artifacts.](https://youtu.be/DPWUcK6mYUw) v2 worked fine for me with this config.

Possible duplicate of https://github.com/yshui/compton/issues/33 but that issue specifically mentioned nouveau so I thought I'd open a new one.

**Comment by yshui** [comment]
Can you confirm whether `backend = "xrender"` works fine?

**Comment by yshui** [comment]
And, if you are able to build from source and run, can you test commit 56f7dd36, and see if that works?

**Comment by jeLee6gi** [comment]
> Can you confirm whether `backend = "xrender"` works fine?

I get no artifacts with my config and `backend = "xrender"`

**Comment by jeLee6gi** [comment]
> And, if you are able to build from source and run, can you test commit [56f7dd3](https://github.com/yshui/compton/commit/56f7dd36f618daf0b38ae37f489e6a3c2d3e1655), and see if that works?

I'm sorry but I'm confused, how do you build https://github.com/yshui/compton/commit/56f7dd36f618daf0b38ae37f489e6a3c2d3e1655? It does not contain meson.build files.

**Comment by yshui** [comment]
@jeLee6gi That commit is before the meson conversion. You build it by running `make`.

**Comment by yshui** [comment]
I have some guesses. Can you record a trace of compton --config /dev/null --backend glx use apitrace? Thanks.

**Comment by yshui** [comment]
And if you can, can you record a trace of compton v2 as well? So we can compare.

**Comment by jeLee6gi** [comment]
I made a mistake earlier, https://github.com/yshui/compton/commit/56f7dd36f618daf0b38ae37f489e6a3c2d3e1655 still has artifacts for me. I don't get flickering but windows are not redrawn. I accidentally ran it with the `xrender` backend before, my bad.

Here are the logs for v2 and v3, I tried to keep them short, I hope they are long enough

[compton_v2.trace.log](https://github.com/yshui/compton/files/2532041/compton_v2.trace.log)
[compton_v3.trace.log](https://github.com/yshui/compton/files/2532042/compton_v3.trace.log)

**Comment by yshui** [comment]
@jeLee6gi As long as the flickering happened when you were capturing the trace, it is good.

**Comment by yshui** [comment]
> I don't get flickering but windows are not redrawn.

That seems to be a different issue. I will take that as 56f7dd3 fixes the flickering, but has another bug (which is likely to have been fixed by now).

**Comment by jeLee6gi** [comment]
Looking at my xorg.conf again was a good lead, I had modified mine according to https://wiki.archlinux.org/index.php/intel_graphics#Tearing and https://wiki.archlinux.org/index.php/intel_graphics#DRI3_issues so it looked like this.

```
Section "Device"
  Identifier    "Intel Graphics"
  Driver        "intel"
  Option        "DRI" "2"
  Option        "TearFree" "true"
EndSection
```

I took out the fallback to DRI 2 and it seems like it fixes the visual glitches and redraw issues with compton v3 on my end. I do see the problems I had with DRI 3 again though, mostly redraw problems when resizing windows but they are only a minor annoyance.

**Comment by yshui** [comment]
Hmmm. So you trace of v2 and v3 looks basically the same. Plus using DRI 2 instead of DRI 3 fixes the problem. I would call this a hardware/driver problem. 

Instead of fiddling with how compton uses opengl to find a workaround, I suggest you rise this issue with intel developers, and see if they have better ideas. (And you should report the DRI 3 issue too)

I will keep this issue open in the meantime.

**Comment by jeLee6gi** [comment]
I agree, it sounds like a driver problem now. Thank you for helping me debug this issue, it works for me and from my side this issue is resolved. If I learn more about this issue I will reply in this thread.

**Comment by tuyenpm9** [comment]
I use integrated intel graphics too, I don't find any issue, no tearing at all.

**Comment by osleg** [comment]
Integrated intel here as well, using modesetting driver instead of intel. No tearing, vsync works like a charm, also i <3 your config :)

**Comment by tuyenpm9** [comment]
@osleg : any benefits of using `modesettings` driver over `intel`?

**Comment by osleg** [comment]
Iirc Intel driver is deprecated long time ago by Intel

**Comment by tuyenpm9** [comment]
@osleg : I have hd630 but enable `modesetting` driver in /etc/x11... causing much tearing than intel, and according to packages page of archlinux, latest update of xf86-video-intel is 26-Oct-2018: https://www.archlinux.org/packages/extra/x86_64/xf86-video-intel/

Don't know what to think, even upsteam stop contribute, it seems opensource project and still something relevant and work for me.

@jeLee6gi : I have to admit, your config is very good.

@yshui : Is Lee config is similar to default? If not how to get default options list of compton?

**Comment by osleg** [comment]
Yeah tearimg happens, there are some keys to fix that but tbh the only thing I use to prevent tearing is compton and it works like a charm 😉

**Comment by yshui** [comment]
@jeLee6gi Hi, can you test the `dri2-hack` branch, see if it works without DRI3?

**Comment by jeLee6gi** [comment]
Your hack fixes the flickering under DRI2 on my end too, thanks for investigating and fixing it so quickly!
